docs/design.md

```diff
@@ -0,0 +1,257 @@
+}
+```
+- `ID` is the unique identifier used by the data provider to identify this update.
+- `Previous` links the current advertisement for the provider with the previous one. This allows
+providers to track the announcements they are broadcasting, and 
+indexer nodes to realize when they have missed a previous advertisement.
+- `Provider` determines the source of the advertisement.
+- `Signature` is a signature over `ID` and `Previous` to authenticate that the advertisement has been generated by `Provider`.
+- `GraphSupport` is used by providers to signal indexer nodes that they support IPLD-aware ingestion sessions (see [protocol](#protocol)).
+
+`AdvertisementID` can be chosen by data providers in the way that suit them best according to their internal
+data indexing strategy. The only requirement for
+`AdvertisementID` is that it should allow providers to uniquely identify the batch of data, or data update
+they are referring to in the advertisement. Indexer nodes use these IDs in their ingestion requests to fetch indexing
+data from the provider.
+
+For instance, a Filecoin miner may choose to use as `AdvertisementID` a combination of a `dealID` with a prefix to identify if
+the advertisement points to a new deal or the expiration of a deal. Thus, when a miner receives an ingestion
+request for `AdvertisementID` from an indexer node, it can easily build the ingestion response (either with the
+batch of new CIDs provided in the deal, or with the list of CIDs the indexer should stop providing due to a deal expiration). Other data providers may choose to generate their `AdvertisementID`s in alternative ways.
+
+For indexer nodes, `AdvertisementID` is an opaque `String` and has no additional significance apart from the unique ID
+for a data provider advertisement.
+
+```go
+type AdvertisementID String
+```
+<!-- We may not need to include `Previous` in Advertisements, as they can be signalled in Ingestion responses. When an advertisement is seen, a request is sent to providers, and providers can signal the previous one. Then the indexer chooses if it is already in sync or not. However, chaining advertisements comes pretty handy for data providers, as it gives them a way to identify the sequence of updates and serve ingestion requests -->
+
+### Ingestion Request
+When an indexer node finds an unseen `Advertisement` and realizes that it is not in sync with a data provider, it sends
+an ingestion request to the data provider to index/remove data.
+```go
+type IngestionRequest struct {
+        ID              AdvertisementID
+        Start           uint64
+        Count           uint64
+}
+```
+- `ID`: `AdvertisementID` for which data is being requested.
+- `Start`: Starting entry from which data wants to be requested.
+- `Count`: Maximum number of entries that want to be received in the response.
+....
+### Ingestion Response
+Data providers use `IngestionResponse`s to answer to ingestion requests.
+
+```go
+type IngestionResponse struct {
```

Comment:
note that unless signature is over hash of overall advertisement there's no authenticity guarantee over the actual cid payloads here,
so a provider could offer an advertisement but map it to different sets of CIDs to different index nodes.